frontend/lib/parsing/parsing-queries.cpp

```diff
@@ -1425,11 +1429,26 @@ const ID& idToParentId(Context* context, ID id) {
   // set this query as an alternative to computing maps
   // in Builder::Result and then redundantly setting them here?
 
+  // Performance: Could this query use id.parentSymbolId in many cases?
+
   ID result;
 
-  const BuilderResult* r = parseFileContainingIdToBuilderResult(context, id);
+  UniqueString parentSymbolPath;
+  const BuilderResult* r =
+    parseFileContainingIdToBuilderResult(context, id, &parentSymbolPath);
+
   if (r != nullptr) {
     result = r->idToParentId(id);
+    // For a submodule in a separate file, the BuilderResult's idToParentId
+    // will return an empty ID for the submodule.
+    // Detect that and return the parent module in that case.
+    if (result.isEmpty() && !parentSymbolPath.isEmpty()) {
+      ID parentSymbolId = id.parentSymbolId(context);
+      if (!parentSymbolId.isEmpty()) {
```

Comment:
Shouldn't `parentSymbold` always be non-empty here, since we know that the current module is nested within another (`!parentSymbolPath.isEmpty()` ?) Is there a situation in which `parentSymbolPath` is non-empty but `parentSymbolId` returns an empty ID?

Reply:
It should not. I can change this into a `CHPL_ASSERT`.

Reply:
I think that would be more robust.